python/ray/tune/util.py

```diff
@@ -2,22 +2,102 @@
 from __future__ import division
 from __future__ import print_function
 
-import logging
 import base64
+import copy
 import fnmatch
+import logging
 import os
-import copy
-import numpy as np
+import threading
 import time
+from collections import defaultdict
+from threading import Thread
 
+import numpy as np
 import ray
 
 logger = logging.getLogger(__name__)
 
+try:
+    import psutil
+except ImportError:
+    psutil = None
+
+try:
+    import GPUtil
+except ImportError:
+    GPUtil = None
+
 _pinned_objects = []
 PINNED_OBJECT_PREFIX = "ray.tune.PinnedObject:"
 
 
+class UtilMonitor(Thread):
+    """Class for system usage utilization monitoring.
+
+    It keeps track of CPU, RAM, GPU, VRAM usage (each gpu separately) by
+    pinging for information every x seconds in a separate thread.
+
+    Requires psutil and GPUtil to be installed. Can be disabled
+    by setting `tune.run(log_sys_usage=False)`.
+    """
+
+    def __init__(self, start=True, delay=0.7):
+        self.stopped = True
+        if GPUtil is None:
+            logger.warning("Install gputil for GPU system monitoring.")
+
+        if psutil is None:
+            logger.warning("Install psutil to monitor system performance.")
+
+        if GPUtil is None and psutil is None:
+            return
+
+        super(UtilMonitor, self).__init__()
+        self.delay = delay  # Time between calls to GPUtil
+        self.values = defaultdict(list)
+        self.lock = threading.Lock()
+        self.daemon = True
+        if start:
+            self.start()
+
+    def _read_utilization(self):
+        with self.lock:
+            if psutil is not None:
+                self.values["cpu"].append(
+                    float(psutil.cpu_percent(interval=None)))
+                self.values["ram"].append(
+                    float(getattr(psutil.virtual_memory(), "percent")))
```

Comment:
Can we be a bit more verbose, "cpu_percent", "ram_percent", and so on? It's hard to tell what the values are otherwise.

Reply:
I've added cpu_util_percent as cpu utilization percentage